src/privatesend-util.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2014-2017 The Dash Core developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#include "privatesend-util.h"
+
+KeyHolder::KeyHolder(CWallet* pwallet)
+        :reserveKey(pwallet)
+{
+        reserveKey.GetReservedKey(pubKey, false);
+}
+
+void KeyHolder::KeepKey()
+{
+    reserveKey.KeepKey();
+}
+
+void KeyHolder::ReturnKey()
+{
+    reserveKey.ReturnKey();
+}
+
+CScript KeyHolder::GetScriptForDestination()
+{
+    return ::GetScriptForDestination(pubKey.GetID());
+}
+
+
+KeyHolderPtr KeyHolderStorage::AddKey(CWallet* pwallet)
+{
+    LogPrintf("PrivateSend - KeyHolderStorage -- AddKey\n");
+    auto key = std::make_shared<KeyHolder>(pwallet);
+    storage.push_back(key);
+    return key;
+}
+
+void KeyHolderStorage::KeepAll(){
+    for(auto key : storage) {
+        key->KeepKey();
+    }
+    storage.clear();
+}
+
+void KeyHolderStorage::ReturnAll()
+{
+    for(auto key : storage) {
+        key->ReturnKey();
+    }
+    storage.clear();
+}
+
+void KeyHolderStorage::ClearOnFailure()
```

Comment:
Looks like logging public vs non-logging private member functions;
`KeepAll` and `ReturnAll` are (non-logging) private helpers,
only used in implementing (logging) `ClearOn*` public methods.

`ClearOnSuccess` & `ClearWhenNotSure` are identical in caling `KeepAll`, just with different log messages.
That is not intuitive; the reader needs to check the definition to see that there's no real difference.

The `Clear` names are less clear to me than `Keep` and `Return`.
Maybe make `KeepAll` and `ReturnAll` as the only two logging public member functions,
and remove the private helpers as they are not used elsewhere.